elease Assert isInTopLayer() in Element::removeFromTopLayer()

anil-hebbar opened the pull request:
#### 02b3689a512e18263e9842b5604069a1cbddfe69
<pre>
elease Assert isInTopLayer() in Element::removeFromTopLayer()
<a href="https://bugs.webkit.org/show_bug.cgi?id=295300">https://bugs.webkit.org/show_bug.cgi?id=295300</a>
<a href="https://rdar.apple.com/149686949">rdar://149686949</a>

Reviewed by Ryosuke Niwa and Tim Nguyen.

The issue is caused by the race condition where the top layer is in
the process of being removed. The fix is to check the element it is
still the top layer before calling the removeFromTopLayer() function

MERGEING WITHOUT TEST REASON
============================
The test for this is not reliable. The test needs lots of reloads to
fail when it should (without the fix) and often timesout if the number
of reloads are more. The EWS failed with errors due to timeout or
texts being dumped when running EWS test. I am therefore merging
without tests

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::hidePopoverInternal):

Canonical link: <a href="https://commits.webkit.org/297056@main">https://commits.webkit.org/297056@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60ee549ac9e657e5893eadf45a6a245367e45a77

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110210 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29869 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20301 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116234 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60458 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112173 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30547 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38456 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83795 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113158 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24373 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99257 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64241 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23737 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17391 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60028 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93750 "Found 1 new API test failure: TestWebKitAPI.GPUProcess.ExitsUnderMemoryPressureWebAudioCase (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17451 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119023 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37250 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27622 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92765 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37622 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95527 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92590 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23632 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37584 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15327 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33139 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37144 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42615 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36806 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40146 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38515 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/fast/dom/remove-top-layer-crash.html:
```diff
@@ -0,0 +1,35 @@
+        window.testRunner?.dumpAsText();
+        let counter = localStorage.getItem('counter');
+        if (!counter)
+        counter = 1;
```
Please indent this line.

anil-hebbar replied:
Done

rniwa commented on LayoutTests/fast/dom/remove-top-layer-crash.html:
```diff
@@ -0,0 +1,35 @@
+        if (!counter)
+        counter = 1;
+        else
+        counter = +counter;
```
Ditto.

anil-hebbar replied:
Done

nt1m commented on LayoutTests/fast/dom/remove-top-layer-crash.html:
```diff
@@ -0,0 +1,35 @@
+            location.reload();
+        } else window.testRunner?.notifyDone();
+    </script>
+    <p> Test PASS if no crash </p>
```
```suggestion
    <p>PASS if test does not crashes</p>
```

The sentence is typically "PASS if test does not crashes"

nt1m reviewed (commented):
missing "R" in the title

anil-hebbar committed:
elease Assert isInTopLayer() in Element::removeFromTopLayer()
https://bugs.webkit.org/show_bug.cgi?id=295300
rdar://149686949

Reviewed by Ryosuke Niwa and Tim Nguyen.

The issue is caused by the race condition where the top layer is in
the process of being removed. The fix is to check the element it is
still the top layer before calling the removeFromTopLayer() function

MERGEING WITHOUT TEST REASON
============================
The test for this is not reliable. The test needs lots of reloads to
fail when it should (without the fix) and often timesout if the number
of reloads are more. The EWS failed with errors due to timeout or
texts being dumped when running EWS test. I am therefore merging
without tests

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::hidePopoverInternal):

Canonical link: https://commits.webkit.org/297056@main
